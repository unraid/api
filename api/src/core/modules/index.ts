// Created from 'create-ts-index'

export * from './array';
export * from './debug';
export * from './disks';
export * from './docker';
export * from './services';
export * from './settings';
export * from './shares';
export * from './users';
export * from './vms';
export * from './add-share';
export * from './add-user';
export * from './get-apps';
export * from './get-devices';
export * from './get-disks';
export * from './get-me';
export * from './get-parity-history';
export * from './get-permissions';
export * from './get-services';
export * from './get-users';
export * from './get-welcome';
