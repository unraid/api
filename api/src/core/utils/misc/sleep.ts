/**
 * Sleep for a certain amount of milliseconds.
 * @param ms How many milliseconds to sleep for.
 */
export const sleep = async (ms: number) =>
    new Promise<void>((resolve) => {
        setTimeout(() => {
            resolve();
        }, ms);
    });
