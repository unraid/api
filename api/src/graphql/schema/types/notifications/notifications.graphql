enum NotificationType {
    UNREAD
    ARCHIVED
    RESTORED
}

input NotificationInput {
    id: ID!
    title: String!
    subject: String!
    description: String
    importance: Importance!
    link: String
    type: NotificationType!
    timestamp: String
}

input NotificationFilter {
    importance: Importance
    type: NotificationType
}

type Mutation {
    sendNotification(notification: NotificationInput!): Notification
}

type Query {
    notifications(filter: NotificationFilter): [Notification!]!
}

type Subscription {
    notificationAdded: Notification!
}

enum Importance {
    ALERT
    INFO
    WARNING
}

type Notification {
    id: ID!
    title: String!
    subject: String!
    description: String!
    importance: Importance!
    link: String
    type: NotificationType!
    """ ISO Timestamp for when the notification occurred """
    timestamp: String
}
