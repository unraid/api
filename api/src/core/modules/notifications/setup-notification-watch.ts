import { join } from 'node:path';

import type { FSWatcher } from 'chokidar';
import { watch } from 'chokidar';

import { getters, store } from '@app/store/index';
import { clearNotification, loadNotification } from '@app/store/modules/notifications';
import { FileLoadStatus } from '@app/store/types';

const handleNotificationAdd = (path: string) => {
    store.dispatch(loadNotification({ path }));
};

const handleNotificationRemove = (path: string) => {
    store.dispatch(clearNotification({ path }));
};

let watcher: FSWatcher | null = null;

export const setupNotificationWatch = async (): Promise<FSWatcher | null> => {
    const { notify, status } = getters.dynamix();
    if (status === FileLoadStatus.LOADED && notify?.path) {
        if (watcher) {
            await watcher.close();
        }
        watcher = watch(join(notify.path, 'unread'), {})
            .on('add', (path) => {
                handleNotificationAdd(path);
            })
            .on('unlink', (path) => {
                handleNotificationRemove(path);
            });

        return watcher;
    }
    return null;
};
