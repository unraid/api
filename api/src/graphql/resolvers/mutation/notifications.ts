/*!
 * Copyright 2021 Lime Technology Inc. All rights reserved.
 * Written by: Alexis Tyler
 */

import { ensurePermission } from '@app/core/utils/permissions/ensure-permission';
import { graphqlLogger } from '@app/core';
import { GraphQLClient } from '@app/mothership/graphql-client';
import { SEND_NOTIFICATION_MUTATION } from '@app/graphql/mothership/mutations';
import { getters } from '@app/store';
import { type Resolvers } from '@app/graphql/generated/api/types';

export const sendNotification: NonNullable<Resolvers['Mutation']>['sendNotification'] = async (_, args: { notification }, context) => {
	const { user } = context;

	// Check permissions
	ensurePermission(user, {
		resource: 'notifications',
		action: 'create',
		possession: 'own',
	});

	const client = GraphQLClient.getInstance();
	// If there's no mothership connection then bail
	if (!client) {
		graphqlLogger.error('Mothership is not working');
		throw new Error('Mothership is down');
	}
	const result = await client.query({ 
		query: SEND_NOTIFICATION_MUTATION, 
		variables: {
			notification: args.notification,
			apiKey: getters.config().remote.apikey,
		}
	});
	graphqlLogger.debug('Query Result from Notifications.ts', result);
	return args.notification;
};
