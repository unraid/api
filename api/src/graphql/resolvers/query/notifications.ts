import { ensurePermission } from '@app/core/utils/permissions/ensure-permission';
import { type QueryResolvers } from '@app/graphql/generated/api/types';
import { getters } from '@app/store/index';

export const notificationsResolver: QueryResolvers['notifications'] = async (
    _,
    { filter: { offset, limit, importance, type } },
    context
) => {
    ensurePermission(context.user, {
        possession: 'any',
        resource: 'notifications',
        action: 'read',
    });

    if (limit > 50) {
        throw new Error('Limit must be less than 50');
    }
    return Object.values(getters.notifications().notifications)
        .filter((notification) => {
            if (
                importance &&
                importance !== notification.importance
            ) {
                return false;
            }
            if (type && type !== notification.type) {
                return false;
            }

            return true;
        })
        .sort(
            (a, b) =>
                new Date(b.timestamp ?? 0).getTime() -
                new Date(a.timestamp ?? 0).getTime()
        )
        .slice(
            offset,
            limit + offset
        );
};
