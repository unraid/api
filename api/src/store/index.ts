import { configureStore } from '@reduxjs/toolkit';

import { listenerMiddleware } from '@app/store/listeners/listener-middleware';
import { cache } from '@app/store/modules/cache';
import { configReducer } from '@app/store/modules/config';
import { docker } from '@app/store/modules/docker';
import { dynamicRemoteAccessReducer } from '@app/store/modules/dynamic-remote-access';
import { dynamix } from '@app/store/modules/dynamix';
import { emhttp } from '@app/store/modules/emhttp';
import { mothership } from '@app/store/modules/minigraph';
import { notificationReducer } from '@app/store/modules/notifications';
import { paths } from '@app/store/modules/paths';
import { registration } from '@app/store/modules/registration';
import { remoteGraphQLReducer } from '@app/store/modules/remote-graphql';
import { upnp } from '@app/store/modules/upnp';

export const store = configureStore({
    reducer: {
        config: configReducer,
        dynamicRemoteAccess: dynamicRemoteAccessReducer,
        minigraph: mothership.reducer,
        paths: paths.reducer,
        emhttp: emhttp.reducer,
        registration: registration.reducer,
        remoteGraphQL: remoteGraphQLReducer,
        notifications: notificationReducer,
        cache: cache.reducer,
        docker: docker.reducer,
        upnp: upnp.reducer,
        dynamix: dynamix.reducer,
    },
    middleware: (getDefaultMiddleware) =>
        getDefaultMiddleware({
            serializableCheck: false,
        }).prepend(listenerMiddleware.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const getters = {
    cache: () => store.getState().cache,
    config: () => store.getState().config,
    docker: () => store.getState().docker,
    dynamicRemoteAccess: () => store.getState().dynamicRemoteAccess,
    dynamix: () => store.getState().dynamix,
    emhttp: () => store.getState().emhttp,
    minigraph: () => store.getState().minigraph,
    notifications: () => store.getState().notifications,
    paths: () => store.getState().paths,
    registration: () => store.getState().registration,
    remoteGraphQL: () => store.getState().remoteGraphQL,
    upnp: () => store.getState().upnp,
};
