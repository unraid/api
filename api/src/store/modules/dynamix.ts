import type { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from '@reduxjs/toolkit';
import { merge } from 'lodash-es';

import { type DynamixConfig } from '@app/core/types/ini';
import { loadDynamixConfigFile } from '@app/store/actions/load-dynamix-config-file';
import { FileLoadStatus } from '@app/store/types';
import { type RecursivePartial } from '@app/types';

export type SliceState = {
    status: FileLoadStatus;
} & DynamixConfig;

export const initialState: Partial<SliceState> = {
    status: FileLoadStatus.UNLOADED,
};

export const dynamix = createSlice({
    name: 'dynamix',
    initialState,
    reducers: {
        updateDynamixConfig(state, action: PayloadAction<RecursivePartial<SliceState>>) {
            return merge(state, action.payload);
        },
    },
    extraReducers(builder) {
        builder.addCase(loadDynamixConfigFile.pending, (state) => {
            state.status = FileLoadStatus.LOADING;
        });

        builder.addCase(loadDynamixConfigFile.fulfilled, (state, action) => {
            merge(state, action.payload, { status: FileLoadStatus.LOADED });
        });

        builder.addCase(loadDynamixConfigFile.rejected, (state, action) => {
            merge(state, action.payload, {
                status: FileLoadStatus.FAILED_LOADING,
            });
        });
    },
});

export const { updateDynamixConfig } = dynamix.actions;
