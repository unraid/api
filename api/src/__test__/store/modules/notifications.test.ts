import { setupNotificationWatch } from '@app/core/modules/notifications/setup-notification-watch';
import { sleep } from '@app/core/utils/misc/sleep';
import { store } from '@app/store/index';
import { loadDynamixConfigFile } from '@app/store/modules/dynamix';
import { expect, test } from 'vitest';

test('loads notifications properly', async () => {
    await store.dispatch(loadDynamixConfigFile());
    const watch = setupNotificationWatch();
    expect(watch).not.toBeNull();
    await sleep(400);
    expect(store.getState().notifications.notifications).toMatchSnapshot();
    await watch?.close();
});
