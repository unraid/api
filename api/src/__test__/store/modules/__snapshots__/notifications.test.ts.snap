// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`loads notifications properly 1`] = `
{
  "/app/dev/notifications/unread/Unraid_Parity_check_1683971161.notify": {
    "description": "Canceled",
    "id": "/app/dev/notifications/unread/Unraid_Parity_check_1683971161.notify",
    "importance": "WARNING",
    "link": undefined,
    "subject": "Notice [UNRAID] - Parity check finished (0 errors)",
    "timestamp": "2023-05-13T09:46:01.000Z",
    "title": "Unraid Parity check",
    "type": "UNREAD",
  },
  "/app/dev/notifications/unread/Unraid_Server_Alert_1683570896.notify": {
    "description": "Power failure on UPS Unraid. Running on batteries.",
    "id": "/app/dev/notifications/unread/Unraid_Server_Alert_1683570896.notify",
    "importance": "ALERT",
    "link": undefined,
    "subject": "UPS Alert",
    "timestamp": "2023-05-08T18:34:56.000Z",
    "title": "Unraid Server Alert",
    "type": "UNREAD",
  },
  "/app/dev/notifications/unread/Unraid_Server_Alert_1683570945.notify": {
    "description": "Communications restored with UPS Unraid",
    "id": "/app/dev/notifications/unread/Unraid_Server_Alert_1683570945.notify",
    "importance": "INFO",
    "link": undefined,
    "subject": "UPS Alert",
    "timestamp": "2023-05-08T18:35:45.000Z",
    "title": "Unraid Server Alert",
    "type": "UNREAD",
  },
  "/app/dev/notifications/unread/Unraid_Server_Alert_1683570980.notify": {
    "description": "Communications restored with UPS Unraid",
    "id": "/app/dev/notifications/unread/Unraid_Server_Alert_1683570980.notify",
    "importance": "INFO",
    "link": undefined,
    "subject": "UPS Alert",
    "timestamp": "2023-05-08T18:36:20.000Z",
    "title": "Unraid Server Alert",
    "type": "UNREAD",
  },
  "/app/dev/notifications/unread/Unraid_Server_Alert_1683571299.notify": {
    "description": "Communications restored with UPS Unraid",
    "id": "/app/dev/notifications/unread/Unraid_Server_Alert_1683571299.notify",
    "importance": "INFO",
    "link": undefined,
    "subject": "UPS Alert",
    "timestamp": "2023-05-08T18:41:39.000Z",
    "title": "Unraid Server Alert",
    "type": "UNREAD",
  },
  "/app/dev/notifications/unread/Unraid_Server_Alert_1683571468.notify": {
    "description": "Communications restored with UPS Unraid",
    "id": "/app/dev/notifications/unread/Unraid_Server_Alert_1683571468.notify",
    "importance": "INFO",
    "link": undefined,
    "subject": "UPS Alert",
    "timestamp": "2023-05-08T18:44:28.000Z",
    "title": "Unraid Server Alert",
    "type": "UNREAD",
  },
  "/app/dev/notifications/unread/Unraid_Server_Alert_1683571535.notify": {
    "description": "Communications restored with UPS Unraid",
    "id": "/app/dev/notifications/unread/Unraid_Server_Alert_1683571535.notify",
    "importance": "INFO",
    "link": undefined,
    "subject": "UPS Alert",
    "timestamp": "2023-05-08T18:45:35.000Z",
    "title": "Unraid Server Alert",
    "type": "UNREAD",
  },
  "/app/dev/notifications/unread/Unraid_Server_Alert_1683571565.notify": {
    "description": "Communications restored with UPS Unraid",
    "id": "/app/dev/notifications/unread/Unraid_Server_Alert_1683571565.notify",
    "importance": "INFO",
    "link": undefined,
    "subject": "UPS Alert",
    "timestamp": "2023-05-08T18:46:05.000Z",
    "title": "Unraid Server Alert",
    "type": "UNREAD",
  },
  "/app/dev/notifications/unread/Unraid_Status_1683260401.notify": {
    "description": "Array has 9 disks (including parity & pools)",
    "id": "/app/dev/notifications/unread/Unraid_Status_1683260401.notify",
    "importance": "INFO",
    "link": undefined,
    "subject": "Notice [UNRAID] - array health report [PASS]",
    "timestamp": "2023-05-05T04:20:01.000Z",
    "title": "Unraid Status",
    "type": "UNREAD",
  },
  "/app/dev/notifications/unread/Unraid_Status_1683346801.notify": {
    "description": "Array has 9 disks (including parity & pools)",
    "id": "/app/dev/notifications/unread/Unraid_Status_1683346801.notify",
    "importance": "INFO",
    "link": undefined,
    "subject": "Notice [UNRAID] - array health report [PASS]",
    "timestamp": "2023-05-06T04:20:01.000Z",
    "title": "Unraid Status",
    "type": "UNREAD",
  },
  "/app/dev/notifications/unread/Unraid_Status_1683433201.notify": {
    "description": "Array has 9 disks (including parity & pools)",
    "id": "/app/dev/notifications/unread/Unraid_Status_1683433201.notify",
    "importance": "INFO",
    "link": undefined,
    "subject": "Notice [UNRAID] - array health report [PASS]",
    "timestamp": "2023-05-07T04:20:01.000Z",
    "title": "Unraid Status",
    "type": "UNREAD",
  },
  "/app/dev/notifications/unread/Unraid_Status_1683519601.notify": {
    "description": "Array has 9 disks (including parity & pools)",
    "id": "/app/dev/notifications/unread/Unraid_Status_1683519601.notify",
    "importance": "INFO",
    "link": undefined,
    "subject": "Notice [UNRAID] - array health report [PASS]",
    "timestamp": "2023-05-08T04:20:01.000Z",
    "title": "Unraid Status",
    "type": "UNREAD",
  },
  "/app/dev/notifications/unread/Unraid_Status_1683606001.notify": {
    "description": "Array has 9 disks (including parity & pools)",
    "id": "/app/dev/notifications/unread/Unraid_Status_1683606001.notify",
    "importance": "INFO",
    "link": undefined,
    "subject": "Notice [UNRAID] - array health report [PASS]",
    "timestamp": "2023-05-09T04:20:01.000Z",
    "title": "Unraid Status",
    "type": "UNREAD",
  },
  "/app/dev/notifications/unread/Unraid_Status_1683692401.notify": {
    "description": "Array has 9 disks (including parity & pools)",
    "id": "/app/dev/notifications/unread/Unraid_Status_1683692401.notify",
    "importance": "INFO",
    "link": undefined,
    "subject": "Notice [UNRAID] - array health report [PASS]",
    "timestamp": "2023-05-10T04:20:01.000Z",
    "title": "Unraid Status",
    "type": "UNREAD",
  },
  "/app/dev/notifications/unread/Unraid_Status_1683778802.notify": {
    "description": "Array has 9 disks (including parity & pools)",
    "id": "/app/dev/notifications/unread/Unraid_Status_1683778802.notify",
    "importance": "INFO",
    "link": undefined,
    "subject": "Notice [UNRAID] - array health report [PASS]",
    "timestamp": "2023-05-11T04:20:02.000Z",
    "title": "Unraid Status",
    "type": "UNREAD",
  },
  "/app/dev/notifications/unread/Unraid_Status_1683865201.notify": {
    "description": "Array has 9 disks (including parity & pools)",
    "id": "/app/dev/notifications/unread/Unraid_Status_1683865201.notify",
    "importance": "INFO",
    "link": undefined,
    "subject": "Notice [UNRAID] - array health report [PASS]",
    "timestamp": "2023-05-12T04:20:01.000Z",
    "title": "Unraid Status",
    "type": "UNREAD",
  },
  "/app/dev/notifications/unread/Unraid_Status_1683951601.notify": {
    "description": "Array has 9 disks (including parity & pools)",
    "id": "/app/dev/notifications/unread/Unraid_Status_1683951601.notify",
    "importance": "INFO",
    "link": undefined,
    "subject": "Notice [UNRAID] - array health report [PASS]",
    "timestamp": "2023-05-13T04:20:01.000Z",
    "title": "Unraid Status",
    "type": "UNREAD",
  },
  "/app/dev/notifications/unread/Unraid_Status_1684038001.notify": {
    "description": "Array has 9 disks (including parity & pools)",
    "id": "/app/dev/notifications/unread/Unraid_Status_1684038001.notify",
    "importance": "INFO",
    "link": undefined,
    "subject": "Notice [UNRAID] - array health report [PASS]",
    "timestamp": "2023-05-14T04:20:01.000Z",
    "title": "Unraid Status",
    "type": "UNREAD",
  },
  "/app/dev/notifications/unread/Unraid_Status_1684124401.notify": {
    "description": "Array has 9 disks (including parity & pools)",
    "id": "/app/dev/notifications/unread/Unraid_Status_1684124401.notify",
    "importance": "INFO",
    "link": undefined,
    "subject": "Notice [UNRAID] - array health report [PASS]",
    "timestamp": "2023-05-15T04:20:01.000Z",
    "title": "Unraid Status",
    "type": "UNREAD",
  },
}
`;
